File: algorithms/2_medium/C/00204_count_primes/global_vars_for_speed.cpp
const size_t MAX_N{ 5'000'001 };
array<bool, MAX_N> is_prime;
array<int, MAX_N> prefix_count;
bool should_init_array{ true };

class Solution {
public:
    Solution() {
        if (should_init_array) {
            for (auto& p : is_prime) {
                p = true;
            }

            for (auto& c : prefix_count) {
                c = 0;
            }

            is_prime[0] = false;
            is_prime[1] = false;

            const size_t ceil_sqrt_max_n{ static_cast<size_t>(ceil(sqrt(MAX_N))) };

            for (size_t i{ 0 }; i <= ceil_sqrt_max_n; ++i) {
                if (is_prime[i]) {
                    for (size_t j{ i * 2 }; j < MAX_N; j += i) {
                        is_prime[j] = false;
                    }
                }
            }
            for (size_t i{ 1 }; i < MAX_N; ++i) {
                prefix_count[i] = prefix_count[i - 1] + (is_prime[i - 1] ? 1 : 0);
            }

            should_init_array = false;
        }
    }

    int countPrimes(int n) {
        return prefix_count[static_cast<size_t>(n)];
    }
};

File: algorithms/2_medium/C/00204_count_primes/global_vars_for_speed.rs
impl Solution {
    pub fn count_primes(n: i32) -> i32 {
        static MAX_N: usize = 5_000_001;
        static mut SHOULD_INIT_ARRAY: bool = true;
        static mut IS_PRIME: [bool; MAX_N] = [true; MAX_N];
        static mut PREFIX_COUNT: [i32; MAX_N] = [0; MAX_N];

        unsafe {
            if SHOULD_INIT_ARRAY {
                IS_PRIME[0] = false;
                IS_PRIME[1] = false;

                let ceil_sqrt_max_n = (MAX_N as f64).sqrt().ceil() as usize;

                for i in 0..=ceil_sqrt_max_n {
                    if IS_PRIME[i] {
                        for j in ((i * 2)..MAX_N).step_by(i) {
                            IS_PRIME[j] = false;
                        }
                    }
                }

                for i in 1..MAX_N {
                    PREFIX_COUNT[i] = PREFIX_COUNT[i - 1] + if IS_PRIME[i - 1] { 1 } else { 0 };
                }

                SHOULD_INIT_ARRAY = false;
            }

            PREFIX_COUNT[n as usize]
        }        
    }
}
